test/integration/auth/node_test.go

```diff
@@ -721,3 +726,391 @@ func expectAllowed(t *testing.T, f func() error) {
+		return func() error {
+			_, err := client.CoreV1().Pods("ns").Create(context.TODO(), &corev1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "pod1",
+					UID:  "uid1",
+				},
+				Spec: corev1.PodSpec{
+					NodeName:           "node1",
+					Containers:         []corev1.Container{{Name: "image", Image: "busybox"}},
+					ServiceAccountName: "default",
+					Volumes:            volumes,
+				},
+			}, metav1.CreateOptions{})
+			return err
+		}
+	}
+
+	getPod := func(client clientset.Interface, podName string) (*corev1.Pod, error) {
+		return client.CoreV1().Pods("ns").Get(context.TODO(), podName, metav1.GetOptions{})
+	}
+
+	deletePod := func(client clientset.Interface, podName string) func() error {
+		return func() error {
+			return client.CoreV1().Pods("ns").Delete(context.TODO(), podName, metav1.DeleteOptions{GracePeriodSeconds: ptr.To[int64](0)})
+		}
+	}
+
+	createCSIDriver := func(client clientset.Interface) func() error {
+		return func() error {
+			_, err := client.StorageV1().CSIDrivers().Create(context.TODO(), &storagev1.CSIDriver{
+				ObjectMeta: metav1.ObjectMeta{Name: "com.example.csi.mydriver"},
+				Spec: storagev1.CSIDriverSpec{
+					TokenRequests: []storagev1.TokenRequest{{Audience: "foo"}},
+				},
+			}, metav1.CreateOptions{})
+			return err
+		}
+	}
+
+	node1Client, _ := clientsetForToken(tokenNode1, clientConfig)
+	expectAllowed(t, createNode1(node1Client))
+	expectAllowed(t, createDefaultServiceAccount(superuserClient))
+
+	projectedVolumeSourceEmptyAudience := &corev1.ProjectedVolumeSource{Sources: []corev1.VolumeProjection{{ServiceAccountToken: &corev1.ServiceAccountTokenProjection{Audience: "", Path: "path"}}}}
+	expectAllowed(t, createPod1(superuserClient, []corev1.Volume{{Name: "foo", VolumeSource: corev1.VolumeSource{Projected: projectedVolumeSourceEmptyAudience}}}))
+	pod, _ := getPod(node1Client, "pod1")
+	expectAllowed(t, createTokenRequestPod1(node1Client, pod.UID, ""))
+	expectAllowed(t, deletePod(superuserClient, "pod1"))
+
+	projectedVolumeSource := &corev1.ProjectedVolumeSource{Sources: []corev1.VolumeProjection{{ServiceAccountToken: &corev1.ServiceAccountTokenProjection{Audience: "foo", Path: "path"}}}}
+	expectAllowed(t, createPod1(superuserClient, []corev1.Volume{{Name: "foo", VolumeSource: corev1.VolumeSource{Projected: projectedVolumeSource}}}))
+	pod, _ = getPod(superuserClient, "pod1")
+	expectAllowed(t, createTokenRequestPod1(superuserClient, pod.UID, "foo"))
+	expectAllowed(t, deletePod(superuserClient, "pod1"))
+
+	csiDriverVolumeSource := &corev1.CSIVolumeSource{Driver: "com.example.csi.mydriver"}
+	expectAllowed(t, createPod1(superuserClient, []corev1.Volume{{Name: "foo", VolumeSource: corev1.VolumeSource{CSI: csiDriverVolumeSource}}}))
+	pod, _ = getPod(superuserClient, "pod1")
+	expectForbidden(t, createTokenRequestPod1(node1Client, pod.UID, "foo"))
+
+	// create csi driver and now token request should be allowed
```

Comment:
Applies to both tests - I think it would probably make sense to run each of these blocks in their own `t.Run` so that this comment can just be the name of the sub test.